lib/hotel_controller.rb

```diff
@@ -0,0 +1,139 @@
+
+      range_length = (end_date - start_date) - 1  #minus 1 so the end date is not counted
+
+      not_available = []   #collects
+
+      range_length.to_i.times do |i|
+        res_by_date_array = reservations(start_date + i)  #a loop that goes over every day in the range and pulls in reservations for that day
+
+        res_by_date_array.each do |res|
+          not_available << res.room      #takes just the room of the reservations for a date and puts them in an array of unavailable rooms
+        end
+      end
+
+      rooms_available_for_range = []
+
+      20.times do |i|
+        if not_available.include?(i+1)   #takes the array made above and looks for each room number 1-20. Note! if the room number shows up just once it is not available. The range may be 5 days but if a room is unavailable for one day it is unavailable for all
+          nil
+        else
+          rooms_available_for_range << (i + 1)  # if the room number is not in the not_available array the room number gets added to the rooms_available array.
+        end
+      end
+
+      if rooms_available_for_range.length > 0
+       return rooms_available_for_range
+      else 
+      raise ArgumentError, "There are no room available for that date range :("
+      end
+    end
+
+    def res_with_valid_dates(range)
+      room = available_rooms(range)[0] #takes first available room
+      
+      reserve_room(range.start_date, range.end_date, room)
+    end
+
+    def block_available(date_range, collection_of_rooms)  #tested in block test because i wrote this for initialization
+      if collection_of_rooms.length < 2 || collection_of_rooms.length > 5
+        raise ArgumentError, "a block needs to have 2 to 5 rooms"
+      end
+
+      available_rooms_array = available_rooms(date_range)
+      collection_of_rooms.each do |room|
+        if available_rooms_array.include?(room)  #checking to make sure each room is available
+          nil
+        else
+          raise ArgumentError, "Room #{room} is not available for that date range :("
+        end
+      end
+      true  #returning true if all the blocks rooms are available.
+    end
+
+    def set_aside_block (date_range, collection_of_rooms, block_id) 
+      unless block_available(date_range, collection_of_rooms)  #if the block_available function returned true do nothing
+        raise ArgumentError, "One or more of your rooms are not available"  # not needed the exceptions is raised in block available
+      end
+
+      collection_of_rooms.each do |room|
+        reserve_room(date_range.start_date, date_range.end_date, room, block_id)
+      end
+    end
+
+    def block_status_by_id(block_id)        #I can check whether a given block has any rooms available
+      @reservation_array.each do |res|
+        if (res.block_id == block_id) && (res.blocks_room_status == nil)
+          return true        #returning true once one block is available
+        else
+           nil
+        end  
```

Comment:
You don't need to have the `else` branch be `nil` here.  That's the default behavior of `if`.

```suggestion
        if (res.block_id == block_id) && (res.blocks_room_status == nil)
          return true        #returning true once one block is available
        end  
```